Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreKit2ObserverModePurchaseDetector.swift
+//
+//  Created by Will Taylor on 5/1/24.
+
+import Foundation
+import StoreKit
+
+/// Protocol describing an actor capable of detecting purchases from StoreKit 2.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol SK2ObserverModePurchaseDetectorType {
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModeManagerDelegate?
+    ) async
+}
+
+/// Actor responsibile for detecting purchases from StoreKit2 that should be processed by observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+actor StoreKit2ObserverModePurchaseDetector: SK2ObserverModePurchaseDetectorType {
+
+    private let deviceCache: DeviceCache
+    private let currentUserProvider: CurrentUserProvider
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        currentUserProvider: CurrentUserProvider,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.currentUserProvider = currentUserProvider
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModeManagerDelegate?
+    ) async {
+        guard let mostRecentVerifiedTransaction: (
+            verifiedTransaction: StoreKit.Transaction,
+            jwsRepresentation: String
+        ) = (await allTransactionsProvider.getAllTransactions()
+            .compactMap { transaction in
+                guard let verifiedTransaction = transaction.verifiedTransaction else {
+                    return nil
+                }
+                return (
+                    verifiedTransaction: verifiedTransaction,
+                    jwsRepresentation: transaction.jwsRepresentation
+                )
+            }
+            .sorted(by: { $0.verifiedTransaction.purchaseDate > $1.verifiedTransaction.purchaseDate })
+            .first
```

Comment:
🤔 how will this get along with consumables? 
A consumable would show up only until its corresponding transaction is finished, so would this mean that we detect twice per consumable (once upon purchase + one false positive after you foreground the app next time after finishing)?

Reply:
We decided to get rid of processing everything returned from Transaction.all to avoid making `n` requests to Khepri for `n` number of transactions the first time the app is opened in observer mode. We figured that if we missed uploading one somehow that Khepri would find it when it queries the list transactions API.

I'm not sure off the top of my head how those consumables would be treated but will do some testing this afternoon! :D

Reply:
Okay, spent some time testing this this afternoon! Here's how it went:

The consumable showed up in `Transaction.all` before it was finished, so it was processed by observer mode before the app's execution resumed from calling `await purchase()`. Then, the app immediately finished the transaction, the app was background/foregrounded again, and it wasn't in Transaction.all since it was finished. Here was the full timeline:

1. User purchases consumable
2. App is background/foregrounded
3. SDK queries `Transaction.all`
4. SDK sploads transaction to Khepri
5. Execution returns to app
6. App finishes the consumable
7. App is background/foregrounded
8. SDK queries `Transaction.all`, but it's empty and nothing is uploaded

Interestingly, when there was an existing subscription when the consumable was purchased, the consumable was in  `Transaction.all` the second time. It was still only processed once though since its ID was cached.

In both tests, the consumable was only processed once and appeared accurately in the RC dashboard

Reply:
For the family sharing option, I _think_ it would probably happen in a similar way where it's all processed before execution flow resumes after `await purchase()`, I'll look at it more tomorrow!

Reply:
> SDK queries Transaction.all, but it's empty and nothing is uploaded

This will only be true when the user has purchased nothing but a single consumable, though, right? 
Like, if I have purchased a subscription, then some time goes by and I purchase a consumable, we will post receipt after the consumable purchase twice: once when the consumable is unfinished, and once again when the consumable is finished, since the latest transaction will be the one for the subscription. 

i.e.: 
1) buy subscription, generate transaction T1
1) foreground app, check history, find transaction T1, send to backend, all is good. Finish transaction. 
1) foreground the app again (like, just open the app). Check history, find transaction T1, not posted to backend because that's the latest one on record. 
1) buy consumable, generate transaction T2
1) foreground app, check history, find transaction T2, send to backend, all is good. Finish transaction. 
1) foreground the app again (like, just open the app). Check history, find transaction T1, post to backend again since the latest one on record is T2 != T1.

Reply:
We decided to discard Mark's suggestion to only cache the most recent transaction ID to avoid this problem. The timeline in your last comment is in line with what I'm seeing in my testing, except for step 6:

Here, we find transaction T1, but we don't post it to the backend again since the cache  is an array of each transaction ID observer mode has previously processed. Since the cache contains the IDs `[T1.id, T2.id]`, observer mode won't ever re-post T1 or T2.

Let me know if that makes sense - I'm happy to hop in a huddle to discuss!